src/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -210,269 +211,245 @@ public sealed partial class TokenCache : ITokenCacheInternal
-                                ea,
-                                requestParams.ClientId,
-                                requestParams.Account?.HomeAccountId?.Identifier,
-                                familyId));
+            IEnumerable<MsalRefreshTokenCacheItem> allRts = Enumerable.Empty<MsalRefreshTokenCacheItem>();
+            await LoadFromCacheAsync(
+               requestParams.RequestContext.TelemetryCorrelationId,
+               CacheEvent.TokenTypes.AT,
+               requestParams.Account,
+               () => allRts = _accessor.GetAllRefreshTokens())
+                   .ConfigureAwait(false);
 
-                    await OnBeforeAccessAsync(args).ConfigureAwait(false);
-                    try
-                    {
-                        // Try to load from all env aliases, but stop at the first valid one
-                        MsalRefreshTokenCacheItem msalRefreshTokenCacheItem = keysAcrossEnvs
-                            .Select(key => _accessor.GetRefreshToken(key))
-                            .FirstOrDefault(item => item != null);
+            // TODO: bogavril - do we want to be silent here?
+            var metadata =
+                await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    allRts.Select(rt => rt.Environment),  // if all environments are known, a network call can be avoided
+                    requestParams.RequestContext)
+                .ConfigureAwait(false);
+            var aliases = metadata.Aliases;
 
-                        requestParams.RequestContext.Logger.Info("Refresh token found in the cache? - " + (msalRefreshTokenCacheItem != null));
+            IEnumerable<MsalRefreshTokenCacheKey> candidateRtKeys = aliases.Select(
+                    al => new MsalRefreshTokenCacheKey(
+                        al,
+                        requestParams.ClientId,
+                        requestParams.Account?.HomeAccountId?.Identifier,
+                        familyId));
 
-                        if (msalRefreshTokenCacheItem != null)
-                        {
-                            return msalRefreshTokenCacheItem;
-                        }
+            MsalRefreshTokenCacheItem candidateRt = allRts.FirstOrDefault(
+                rt => candidateRtKeys.Any(
+                    candidateKey => object.Equals(rt.GetKey(), candidateKey)));
 
-                        requestParams.RequestContext.Logger.Info("Checking ADAL cache for matching RT");
+            requestParams.RequestContext.Logger.Info("Refresh token found in the cache? - " + (candidateRt != null));
 
-                        string upn = string.IsNullOrWhiteSpace(requestParams.LoginHint)
-                            ? requestParams.Account?.Username
-                            : requestParams.LoginHint;
+            if (candidateRt != null)
+                return candidateRt;
```

Comment:
>Rt [](start = 32, length = 2)

nit: i get the acronyms throughout here, but at first glance, they are a little tricky to read...like "allRts"...but maybe it's just me :)

Reply:
No, it's a valid comment, I will update.